### ISSUE TITLE
Compiler Warnings for Visibility Decorators

### ISSUE THREAD

**Issue by pcaversaccio** [vyperlang/vyper#3743]
In terms of devex, having the compiler spit out a warning (or error?) when a function can be declared `view` or `pure` would be valuable I think

### Example

```vy
@external
@view
def uint256_average(x: uint256, y: uint256) -> uint256:
    return unsafe_add(x & y, (x ^ y) >> 1)
```

The compiler could issue something like:

```console
Warning: Function uint256_average state mutability can be restricted to pure
```

Opinions?

**Comment by charles-cooper** [comment]
yes, this sounds good

**Comment by charles-cooper** [comment]
i don't think we need to spit a warning for `pure` vs `view`. but we could raise an error if the function is marked `>= nonpayable` but doesn't touch state.

**Comment by pcaversaccio** [comment]
> i don't think we need to spit a warning for `pure` vs `view`. but we could raise an error if the function is marked `>= nonpayable` but doesn't touch state.

We (=the Vyper compiler) should optimise for maximum correctness. Declaring a `pure` function as a `view` is not wrong per se, but it is still not maximally correct and we should help the user to understand this. So a short warning helps to understand what can be improved and does not harm anyone.

Regarding the second point: what's the reason to raise here instead of also issuing a warning? It's again not wrong per se (because you simply declare a function to be more powerful than it is, while the function logic does not even use this power), just not maximally correct? I would go for warnings everywhere.

**Comment by charles-cooper** [comment]
in general i think warnings are pretty useless except for deprecating features. essentially, if it's worth emitting a warning, it's worth emitting an error. (e.g. i think the best practice for compilers like `gcc` which issue many kinds of warnings, is to use `-Werror` to convert warnings to errors).

>  Declaring a pure function as a view is not wrong per se, but it is still not maximally correct and we should help the user to understand this. So a short warning helps to understand what can be improved and does not harm anyone.

there isn't currently that much semantic difference in vyper between pure and view. they behave identically wrt to optimization, and the difference at this point is more pedantic than anything. i guess it's a useful marker sometimes, if you see that a function is `@pure`, you know it doesn't touch the environment. so i guess i'm fairly neutral on the pure vs view issue.

**Comment by pcaversaccio** [comment]
> if it's worth emitting a warning, it's worth emitting an error.

There is a semantic difference between a warning and an error:
- An error indicates a violation of the programming language's syntax rules or semantic rules.
- A warning, on the other hand, indicates potential issues or questionable practices in the code but does not prevent the compilation process from completing.

We can go for errors for functions that are marked as state-modifying but don't modify (but maybe read) any state, even though I would consider this rather being overzealous. For the `view` vs `pure`, it's really just about having a warning to indicate what _would_ be maximally correct. I'm fine having a linter doing this as well, but this is definitely "questionable practice" if you mark a `pure` function as `view`. Also, see https://github.com/vyperlang/vyper/issues/3093 and https://github.com/vyperlang/vyper/issues/3141 for issues where you can touch state in `pure` functions.
